File: dll/src/text_session.cpp
#include "text_session.h"
#include "../../include/message_event.h"
#include "../../include/connector_socket.h"
#include "../../include/fxnet_interface.h"
#include "../../include/log_utility.h"
#include "../../include/net_stream_package.h"
#include "../../utility/time_utility.h"

#include <string>
#include <stdlib.h>
#include <stdio.h>

CTextSession::TextMessageEvent::TextMessageEvent(ISession* pSession)
	: m_pSession(pSession)
{
}

void CTextSession::TextMessageEvent::operator()(std::ostream* pOStream)
{
	DELETE_WHEN_DESTRUCT(TextMessageEvent, this);

	this->m_pSession->OnRecv(this->m_oPackage, pOStream);
}

CTextSession::ConnectedEvent::ConnectedEvent(ISession* pSession)
	: m_pSession(pSession)
{
}

void CTextSession::ConnectedEvent::operator()(std::ostream* pOStream)
{
	DELETE_WHEN_DESTRUCT(ConnectedEvent, this);
	this->m_pSession->OnConnected(pOStream);
}

CTextSession::SessionErrorEvent::SessionErrorEvent(ISession* pSession, const ErrorCode& oError)
	: m_oError(oError)
	, m_pSession(pSession)
{
}

void CTextSession::SessionErrorEvent::operator()(std::ostream* pOStream)
{
	DELETE_WHEN_DESTRUCT(SessionErrorEvent, this);
	this->m_pSession->OnError(this->m_oError, pOStream);
}

CTextSession::CloseSessionEvent::CloseSessionEvent(ISession* pSession)
	: m_pSession(pSession)
{
}

void CTextSession::CloseSessionEvent::operator()(std::ostream* pOStream)
{
	DELETE_WHEN_DESTRUCT(CloseSessionEvent, this);
	this->m_pSession->OnClose(pOStream);
}

CTextSession::SessionOnSendEvent::SessionOnSendEvent(ISession* pSession)
	: m_pSession(pSession)
	, m_dwLen(0)
{
}

void CTextSession::SessionOnSendEvent::operator()(std::ostream* pOStream)
{
	DELETE_WHEN_DESTRUCT(SessionOnSendEvent, this);
}

CTextSession& CTextSession::Send(const char* szData, unsigned int dwLen, std::ostream* pOStream)
{
	class SendOperator : public IOEventBase
	{
	public:
		SendOperator(FXNET::CConnectorSocket* opSock)
			: m_opSock(opSock)
		{
		}
		virtual void operator ()(std::ostream* pOStream)
		{
			DELETE_WHEN_DESTRUCT(SendOperator, this);

			FXNET::CConnectorSocket* poConnector = this->m_opSock;
			if (poConnector->GetError())
			{
				return;
			}

			LOG(pOStream, ELOG_LEVEL_DEBUG4) << this->m_opSock->Name()
				<< ", " << this->m_opSock->NativeSocket()
				<< "\n";
			
			poConnector->GetSession()->GetSendBuff().PushData(m_oPackage);
			ErrorCode oError;
			poConnector->SendMessage(oError, pOStream);
			LOG(pOStream, ELOG_LEVEL_DEBUG4) << this->m_opSock->Name()
				<< ", " << this->m_opSock->NativeSocket()
				<< "\n";
		}
		FXNET::CConnectorSocket* m_opSock;
		FXNET::CNetStreamPackage m_oPackage;
	protected:
	private:
	};

	SendOperator* pOperator = new SendOperator(this->m_opSock);
	pOperator->m_oPackage.WriteString(szData, dwLen);
	FXNET::PostEvent(this->m_opSock->GetIOModuleIndex(), pOperator);
	return *this;
}

CTextSession& CTextSession::OnRecv(FXNET::CNetStreamPackage& refPackage, std::ostream* pOStream)
{
	int dwMagicNum = 0;
	std::string szData;
	this->m_pOnRecv(m_pConnector, refPackage.GetData(), refPackage.GetDataLength());

	return *this;
}

void CTextSession::OnConnected(std::ostream* pOStream)
{
	LOG(pOStream, ELOG_LEVEL_INFO) << GetSocket()->NativeSocket() << ", connected!!!"
		<< "\n";

	this->m_pOnConnected(m_pConnector);
}

void CTextSession::OnError(const ErrorCode& refError, std::ostream* pOStream)
{
	LOG(pOStream, ELOG_LEVEL_ERROR) << this->GetSocket()->NativeSocket() << ", error: " << refError.What()
		<< "\n";

	class ErrorOperator : public IOEventBase
	{
	public:
		ErrorOperator(FXNET::ISocketBase* opSock, ErrorCode dwError)
			: m_opSock(opSock)
			, m_oError(dwError)
		{
		}
		virtual void operator ()(std::ostream* pOStream)
		{
			DELETE_WHEN_DESTRUCT(ErrorOperator, this);

			this->m_opSock->OnError(m_oError, pOStream);
		}
		FXNET::ISocketBase* m_opSock;
		ErrorCode m_oError;
	protected:
	private:
	};

	ErrorOperator* pOperator = new ErrorOperator(m_opSock, refError);
	FXNET::PostEvent(this->m_opSock->GetIOModuleIndex(), pOperator);
	this->m_pOnError(m_pConnector, refError);
}

void CTextSession::OnClose(std::ostream* pOStream)
{
	LOG(pOStream, ELOG_LEVEL_INFO) << this->GetSocket()->NativeSocket()
		<< "\n";

	class OnCloseOperator : public IOEventBase
	{
	public:
		OnCloseOperator(FXNET::ISocketBase* opSock)
			: m_opSock(opSock)
		{
		}
		virtual void operator ()(std::ostream* pOStream)
		{
			DELETE_WHEN_DESTRUCT(OnCloseOperator, this);

			this->m_opSock->OnClose(pOStream);
		}
		FXNET::ISocketBase* m_opSock;
		std::string m_szData;
	protected:
	private:
	};

	OnCloseOperator* pOperator = new OnCloseOperator(m_opSock);
	FXNET::PostEvent(this->m_opSock->GetIOModuleIndex(), pOperator);

	LOG(pOStream, ELOG_LEVEL_INFO) << "session close, " << this
		<< "\n";

	this->SetSock(NULL);

	m_pOnClose(m_pConnector);
}

void CTextSession::Close(std::ostream* pOStream)
{
	class CloseOperator : public IOEventBase
	{
	public:
		CloseOperator(FXNET::ISocketBase* opSock)
			: m_opSock(opSock)
		{
		}
		virtual void operator ()(std::ostream* pOStream)
		{
			DELETE_WHEN_DESTRUCT(CloseOperator, this);

			this->m_opSock->Close(pOStream);
		}
		FXNET::ISocketBase* m_opSock;
	protected:
	private:
	};

	CloseOperator* pOperator = new CloseOperator(m_opSock);
	FXNET::PostEvent(this->m_opSock->GetIOModuleIndex(), pOperator);
}

CTextSession::TextMessageEvent* CTextSession::NewRecvMessageEvent()
{
	CTextSession::TextMessageEvent* pEvent = new CTextSession::TextMessageEvent(this);
	return pEvent;
}

MessageEventBase* CTextSession::NewConnectedEvent()
{
	CTextSession::ConnectedEvent* pEvent = new CTextSession::ConnectedEvent(this);
	return pEvent;
}

CTextSession::SessionErrorEvent* CTextSession::NewErrorEvent(const ErrorCode& refError)
{
	CTextSession::SessionErrorEvent* pEvent = new CTextSession::SessionErrorEvent(this, refError);
	return pEvent;
}

CTextSession::CloseSessionEvent* CTextSession::NewCloseEvent()
{
	CTextSession::CloseSessionEvent* pEvent = new CTextSession::CloseSessionEvent(this);
	return pEvent;
}

CTextSession::SessionOnSendEvent* CTextSession::NewOnSendEvent(int dwLen)
{
	CTextSession::SessionOnSendEvent* pEvent = new CTextSession::SessionOnSendEvent(this);
	pEvent->m_dwLen = dwLen;
	return pEvent;
}

File: network/Packages/com.fxnet.network/Runtime/Dll/TextSession.cs
using System;
using System.Buffers.Binary;
using System.IO;
using FxNet.Core;
using FxNet.IO;

namespace FxNet.Dll
{
    /// <summary>
    /// 文本协议会话实现，采用 4 字节大端长度头 + 数据体的消息格式。
    /// 实现 ISession 接口，负责数据的序列化/反序列化和事件生成。
    /// Send/Close 均通过 PostEvent 投递到 IO 线程执行，与 C++ CTextSession 对齐。
    /// </summary>
    public class TextSession : ISession
    {
        private ConnectorSocket? _socket; // 底层连接器
        private readonly TextWorkStream _sendBuff = new TextWorkStream(); // 发送缓冲区
        private readonly TextWorkStream _recvBuff = new TextWorkStream(); // 接收缓冲区

        private readonly Connector _connector;
        private readonly OnRecvCallback _onRecv;
        private readonly OnConnectedCallback _onConnected;
        private readonly OnErrorCallback _onError;
        private readonly OnCloseCallback _onClose;

        public TextSession(Connector connector,
            OnRecvCallback onRecv, OnConnectedCallback onConnected,
            OnErrorCallback onError, OnCloseCallback onClose)
        {
            _connector = connector;
            _onRecv = onRecv;
            _onConnected = onConnected;
            _onError = onError;
            _onClose = onClose;
        }

        public void SetSocket(ConnectorSocket? socket) => _socket = socket;
        public ConnectorSocket? GetSocket() => _socket;

        /// <summary>
        /// 发送数据：创建 SendOperator 投递到 IO 线程，在 IO 线程中写入发送缓冲区并触发底层发送。
        /// 对齐 C++ CTextSession::Send（创建 SendOperator → PostEvent → IO 线程执行 PushData + SendMessage）。
        /// </summary>
        public ISession Send(byte[] data, int len, TextWriter? output)
        {
            if (_socket == null) return this;
            var op = new SendOperator(_socket);
            op.Package.WriteData(data, len);
            FxNetInterface.PostEvent(_socket.GetIOModuleIndex(), op);
            return this;
        }

        /// <summary>
        /// 关闭连接：创建 CloseOperator 投递到 IO 线程执行底层 Socket 关闭。
        /// 对齐 C++ CTextSession::Close（创建 CloseOperator → PostEvent → IO 线程执行 socket.Close）。
        /// </summary>
        public void Close(TextWriter? output)
        {
            if (_socket == null) return;
            var op = new CloseOperator(_socket);
            FxNetInterface.PostEvent(_socket.GetIOModuleIndex(), op);
        }

        public ISession OnRecv(NetStreamPackage package, TextWriter? output)
        {
            _onRecv(_connector, package.GetData(), package.DataLength);
            return this;
        }

        public void OnConnected(TextWriter? output)
        {
            _onConnected(_connector);
        }

        public void OnError(ErrorCode error, TextWriter? output)
        {
            _onError(_connector, error.Code);
        }

        public void OnClose(TextWriter? output)
        {
            _onClose(_connector);
        }

        public NetWorkStream GetSendBuff() => _sendBuff;
        public NetWorkStream GetRecvBuff() => _recvBuff;

        /// <summary>创建各类跨线程消息事件，由 IO 线程投递并执行</summary>

        public MessageRecvEventBase NewRecvMessageEvent()
        {
            return new TextMessageEvent(this);
        }

        public MessageEventBase NewConnectedEvent()
        {
            return new ConnectedEvent(this);
        }

        public MessageEventBase NewErrorEvent(ErrorCode error)
        {
            return new SessionErrorEvent(this, error);
        }

        public MessageEventBase NewCloseEvent()
        {
            return new CloseSessionEvent(this);
        }

        public MessageEventBase NewOnSendEvent(int len)
        {
            return new SessionOnSendEvent(this, len);
        }

        // === 内部 IO 事件类，用于向 IO 线程投递 Send/Close 操作 ===

        /// <summary>发送操作事件，在 IO 线程中将数据写入发送缓冲区并触发底层发送（对齐 C++ SendOperator）</summary>
        private class SendOperator : IOEventBase
        {
            private readonly ConnectorSocket _socket;
            public NetStreamPackage Package { get; } = new NetStreamPackage();

            public SendOperator(ConnectorSocket socket)
            {
                _socket = socket;
            }

            public override void Execute(TextWriter? output)
            {
                if (_socket.GetError() != 0) return;

                // 写入 4 字节大端长度头 + 数据体（与 TextWorkStream 协议一致）
                var sendBuff = _socket.GetSession()!.GetSendBuff();
                int dataLen = Package.DataLength;
                Span<byte> header = stackalloc byte[4];
                BinaryPrimitives.WriteInt32BigEndian(header, dataLen);
                sendBuff.PushData(header);
                sendBuff.PushData(Package.GetData(), dataLen);
                var error = new ErrorCode();
                _socket.SendMessage(error, output);
            }
        }

        /// <summary>关闭操作事件，在 IO 线程中执行底层 Socket 关闭（对齐 C++ CloseOperator）</summary>
        private class CloseOperator : IOEventBase
        {
            private readonly ConnectorSocket _socket;

            public CloseOperator(ConnectorSocket socket)
            {
                _socket = socket;
            }

            public override void Execute(TextWriter? output)
            {
                _socket.Close(output);
            }
        }

        // === 内部消息事件类，用于跨线程投递 ===

        /// <summary>数据接收事件，在 IO 线程中调用 Session.OnRecv</summary>
        private class TextMessageEvent : MessageRecvEventBase
        {
            private readonly ISession _session;

            public TextMessageEvent(ISession session)
            {
                _session = session;
                Session = session;
            }

            public override void Execute(TextWriter? output)
            {
                _session.OnRecv(Package, output);
            }
        }

        /// <summary>连接建立事件</summary>
        private class ConnectedEvent : MessageEventBase
        {
            private readonly ISession _session;
            public ConnectedEvent(ISession session) => _session = session;
            public override void Execute(TextWriter? output) => _session.OnConnected(output);
        }

        /// <summary>会话错误事件</summary>
        private class SessionErrorEvent : MessageEventBase
        {
            private readonly ISession _session;
            private readonly ErrorCode _error;
            public SessionErrorEvent(ISession session, ErrorCode error)
            {
                _session = session;
                _error = error;
            }
            public override void Execute(TextWriter? output) => _session.OnError(_error, output);
        }

        /// <summary>会话关闭事件</summary>
        private class CloseSessionEvent : MessageEventBase
        {
            private readonly ISession _session;
            public CloseSessionEvent(ISession session) => _session = session;
            public override void Execute(TextWriter? output) => _session.OnClose(output);
        }

        /// <summary>发送完成通知事件</summary>
        private class SessionOnSendEvent : MessageEventBase
        {
            private readonly ISession _session;
            private readonly int _len;
            public SessionOnSendEvent(ISession session, int len)
            {
                _session = session;
                _len = len;
            }
            public override void Execute(TextWriter? output)
            {
                // OnSend notification - could be extended
            }
        }
    }
}
